File: courses/dropbox/drills/typescript/drill_06_search_index.ts
/**
 * Drill 06 — Search Index
 * ========================
 * Dash universal search — build an inverted index that supports keyword
 * search across documents from multiple sources.
 *
 * Level 1 — Basic Indexing (10 min)
 * ----------------------------------
 *   addDocument(docId: string, content: string, source?: string): void
 *       Index a document. Content is a space-separated string of
 *       words. If docId already exists, replace its content.
 *
 *   search(keyword: string): string[]
 *       Return docIds containing the keyword (case-insensitive).
 *       Sorted alphabetically.
 *
 *   removeDocument(docId: string): boolean
 *       Remove from index. Return false if not found.
 *
 * Level 2 — Multi-Keyword Search (10 min)
 * -----------------------------------------
 *   searchAll(keywords: string[]): string[]
 *       Return docIds containing ALL keywords (AND).
 *       Case-insensitive. Sorted alphabetically.
 *
 *   searchAny(keywords: string[]): string[]
 *       Return docIds containing ANY keyword (OR).
 *       Case-insensitive. Sorted alphabetically.
 *
 *   getWordCount(docId: string): number
 *       Number of unique words in the document. -1 if not found.
 *
 * Level 3 — Ranked Results (10 min)
 * -----------------------------------
 *   searchRanked(keyword: string): string[]
 *       Return docIds sorted by relevance: number of times the
 *       keyword appears in the document (descending), then docId
 *       alphabetically for ties. Case-insensitive.
 *
 *   addDocument now accepts an optional source parameter
 *       (e.g. "gmail", "slack", "gdrive"). Default is "unknown".
 *
 *   searchBySource(keyword: string, source: string): string[]
 *       Search filtered to a specific source. Sorted alphabetically.
 *
 * Level 4 — Prefix Search / Autocomplete (15 min)
 * -------------------------------------------------
 *   autocomplete(prefix: string, limit?: number): string[]
 *       Return unique words in the index that start with prefix
 *       (case-insensitive). Sorted alphabetically. Limited to
 *       `limit` results (default 5).
 *
 *   getIndexStats(): { documents: number; unique_words: number; sources: string[] }
 *       where sources is sorted alphabetically.
 *
 * Examples
 * --------
 *   const idx = new SearchIndex();
 *   idx.addDocument("d1", "hello world hello", "gmail");
 *   idx.search("hello");           // -> ["d1"]
 *   idx.searchRanked("hello");     // -> ["d1"]  (2 occurrences)
 *   idx.autocomplete("hel");       // -> ["hello"]
 */


class SearchIndex {
  constructor() {
    // TODO: initialize your data structures
  }

  // -- Level 1 --------------------------------------------------

  /** Index a document. Replace if docId already exists. */
  addDocument(docId: string, content: string, source: string = "unknown"): void {
    throw new Error("TODO: addDocument");
  }

  /** Return docIds containing keyword (case-insensitive), sorted. */
  search(keyword: string): string[] {
    throw new Error("TODO: search");
  }

  /** Remove from index. Return false if not found. */
  removeDocument(docId: string): boolean {
    throw new Error("TODO: removeDocument");
  }

  // -- Level 2 --------------------------------------------------

  /** Return docIds containing ALL keywords (AND). Sorted. */
  searchAll(keywords: string[]): string[] {
    throw new Error("TODO: searchAll");
  }

  /** Return docIds containing ANY keyword (OR). Sorted. */
  searchAny(keywords: string[]): string[] {
    throw new Error("TODO: searchAny");
  }

  /** Number of unique words in the document. -1 if not found. */
  getWordCount(docId: string): number {
    throw new Error("TODO: getWordCount");
  }

  // -- Level 3 --------------------------------------------------

  /** DocIds sorted by keyword frequency desc, then id asc. */
  searchRanked(keyword: string): string[] {
    throw new Error("TODO: searchRanked");
  }

  /** Search filtered to a specific source. Sorted alphabetically. */
  searchBySource(keyword: string, source: string): string[] {
    throw new Error("TODO: searchBySource");
  }

  // -- Level 4 --------------------------------------------------

  /** Words starting with prefix (case-insensitive), sorted, limited. */
  autocomplete(prefix: string, limit: number = 5): string[] {
    throw new Error("TODO: autocomplete");
  }

  /** Return {documents, unique_words, sources}. */
  getIndexStats(): { documents: number; unique_words: number; sources: string[] } {
    throw new Error("TODO: getIndexStats");
  }
}


// --- Self-Checks (do not edit below this line) ------------------

let _passed = 0;
let _failed = 0;

function _check(label: string, actual: unknown, expected: unknown): void {
  if (JSON.stringify(actual) === JSON.stringify(expected)) {
    _passed++;
    console.log(`  \u2713 ${label}`);
  } else {
    _failed++;
    console.log(`  \u2717 ${label}`);
    console.log(`    expected: ${JSON.stringify(expected)}`);
    console.log(`         got: ${JSON.stringify(actual)}`);
  }
}

function _level(name: string, fn: () => void): void {
  console.log(name);
  try {
    fn();
  } catch (e: any) {
    if (e.message?.startsWith("TODO")) {
      console.log(`  \u25cb ${e.message}`);
    } else {
      _failed++;
      console.log(`  \u2717 ${e.message}`);
    }
  }
}

function _runSelfChecks(): void {

  function level1(): void {
    const idx = new SearchIndex();
    // Basic add and search
    idx.addDocument("doc1", "the quick brown fox");
    idx.addDocument("doc2", "the lazy brown dog");
    idx.addDocument("doc3", "quick fox jumps high");
    _check("search 'brown'",
      idx.search("brown"), ["doc1", "doc2"]);
    _check("search 'quick'",
      idx.search("quick"), ["doc1", "doc3"]);

    // Case-insensitive search
    idx.addDocument("doc4", "Hello World HELLO");
    _check("case-insensitive search 'hello'",
      idx.search("hello"), ["doc4"]);
    _check("case-insensitive search 'HELLO'",
      idx.search("HELLO"), ["doc4"]);

    // Search miss
    _check("search miss",
      idx.search("zebra"), []);

    // Replace existing document
    idx.addDocument("doc1", "completely new content");
    _check("replaced doc no longer matches 'fox'",
      idx.search("fox"), ["doc3"]);
    _check("replaced doc matches 'completely'",
      idx.search("completely"), ["doc1"]);

    // Remove
    _check("remove existing returns true",
      idx.removeDocument("doc2"), true);
    _check("removed doc gone from search",
      idx.search("lazy"), []);
    _check("remove missing returns false",
      idx.removeDocument("doc999"), false);
  }

  function level2(): void {
    const idx = new SearchIndex();
    idx.addDocument("d1", "python java rust");
    idx.addDocument("d2", "python rust go");
    idx.addDocument("d3", "java go elixir");
    idx.addDocument("d4", "python java go rust elixir");

    // searchAll — AND
    _check("searchAll ['python', 'rust']",
      idx.searchAll(["python", "rust"]),
      ["d1", "d2", "d4"]);
    _check("searchAll ['java', 'go', 'elixir']",
      idx.searchAll(["java", "go", "elixir"]),
      ["d3", "d4"]);
    _check("searchAll no match",
      idx.searchAll(["python", "elixir", "haskell"]),
      []);

    // searchAny — OR
    _check("searchAny ['elixir']",
      idx.searchAny(["elixir"]),
      ["d3", "d4"]);
    _check("searchAny ['python', 'elixir']",
      idx.searchAny(["python", "elixir"]),
      ["d1", "d2", "d3", "d4"]);
    _check("searchAny no match",
      idx.searchAny(["haskell", "cobol"]),
      []);

    // getWordCount
    _check("word count d1", idx.getWordCount("d1"), 3);
    _check("word count d4", idx.getWordCount("d4"), 5);
    _check("word count missing", idx.getWordCount("d99"), -1);

    // Duplicate words — unique count
    idx.addDocument("d5", "go go go python python");
    _check("word count with dupes", idx.getWordCount("d5"), 2);
  }

  function level3(): void {
    const idx = new SearchIndex();
    // Frequency ranking: "data" appears 3x in d1, 1x in d2, 2x in d3
    idx.addDocument("d1", "data data data science", "gmail");
    idx.addDocument("d2", "data engineering", "slack");
    idx.addDocument("d3", "big data data pipeline", "gmail");
    idx.addDocument("d4", "machine learning model", "gdrive");

    _check("searchRanked by frequency",
      idx.searchRanked("data"),
      ["d1", "d3", "d2"]); // 3, 2, 1 occurrences

    // Tie-breaking by docId alphabetically
    idx.addDocument("a_doc", "data results", "slack");
    idx.addDocument("z_doc", "data results", "slack");
    const ranked = idx.searchRanked("data");
    // a_doc and z_doc both have 1 occurrence, same as d2
    _check("searchRanked tie-breaking",
      ranked,
      ["d1", "d3", "a_doc", "d2", "z_doc"]);

    _check("searchRanked no match",
      idx.searchRanked("quantum"), []);

    // searchBySource
    _check("searchBySource gmail 'data'",
      idx.searchBySource("data", "gmail"),
      ["d1", "d3"]);
    _check("searchBySource slack 'data'",
      idx.searchBySource("data", "slack"),
      ["a_doc", "z_doc"]);
    _check("searchBySource gdrive 'data'",
      idx.searchBySource("data", "gdrive"),
      []);

    // Case-insensitive ranked
    idx.addDocument("d5", "Data DATA data", "gmail");
    _check("ranked case-insensitive -- d5 has 3 occurrences",
      idx.searchRanked("DATA").slice(0, 2),
      ["d1", "d5"]); // both have 3
  }

  function level4(): void {
    const idx = new SearchIndex();
    idx.addDocument("d1", "dropbox desktop application sync", "gdrive");
    idx.addDocument("d2", "dropbox drive document sharing", "gmail");
    idx.addDocument("d3", "application programming interface api", "slack");

    // Autocomplete basic
    _check("autocomplete 'drop'",
      idx.autocomplete("drop"),
      ["dropbox"]);
    _check("autocomplete 'dr'",
      idx.autocomplete("dr"),
      ["drive", "dropbox"]);
    _check("autocomplete 'app'",
      idx.autocomplete("app"),
      ["application"]);

    // Autocomplete case-insensitive
    _check("autocomplete 'D' case-insensitive",
      idx.autocomplete("D"),
      ["desktop", "document", "drive", "dropbox"]);

    // Autocomplete with limit
    idx.addDocument("d4", "alpha beta gamma delta epsilon zeta");
    _check("autocomplete limit 3",
      idx.autocomplete("", 3),
      ["alpha", "api", "application"]);

    // Autocomplete no match
    _check("autocomplete no match",
      idx.autocomplete("xyz"),
      []);

    // getIndexStats
    const stats = idx.getIndexStats();
    _check("stats documents count",
      stats["documents"], 4);
    _check("stats unique_words > 0",
      stats["unique_words"] > 0, true);
    _check("stats sources sorted",
      stats["sources"],
      ["gdrive", "gmail", "slack"]);

    // Remove a doc and check stats update
    idx.removeDocument("d3");
    const stats2 = idx.getIndexStats();
    _check("stats after remove -- documents count",
      stats2["documents"], 3);
  }

  _level("Level 1 \u2014 Basic Indexing", level1);
  _level("Level 2 \u2014 Multi-Keyword Search", level2);
  _level("Level 3 \u2014 Ranked Results", level3);
  _level("Level 4 \u2014 Prefix Search / Autocomplete", level4);
}

console.log("\nDrill 06 \u2014 Search Index\n");
_runSelfChecks();
const _total = _passed + _failed;
console.log(`\n${_passed}/${_total} passed`);
if (_failed === 0 && _total > 0) {
  console.log("All tests passed.");
}

File: courses/dropbox/drills/python/drill_06_search_index.py
"""
Drill 06 — Search Index
========================
Dash universal search — build an inverted index that supports keyword
search across documents from multiple sources.

Level 1 — Basic Indexing (10 min)
----------------------------------
  add_document(doc_id: str, content: str) -> None
      Index a document. Content is a space-separated string of
      words. If doc_id already exists, replace its content.

  search(keyword: str) -> list[str]
      Return doc_ids containing the keyword (case-insensitive).
      Sorted alphabetically.

  remove_document(doc_id: str) -> bool
      Remove from index. Return False if not found.

Level 2 — Multi-Keyword Search (10 min)
-----------------------------------------
  search_all(keywords: list[str]) -> list[str]
      Return doc_ids containing ALL keywords (AND).
      Case-insensitive. Sorted alphabetically.

  search_any(keywords: list[str]) -> list[str]
      Return doc_ids containing ANY keyword (OR).
      Case-insensitive. Sorted alphabetically.

  get_word_count(doc_id: str) -> int
      Number of unique words in the document. -1 if not found.

Level 3 — Ranked Results (10 min)
-----------------------------------
  search_ranked(keyword: str) -> list[str]
      Return doc_ids sorted by relevance: number of times the
      keyword appears in the document (descending), then doc_id
      alphabetically for ties. Case-insensitive.

  add_document now accepts an optional source: str parameter
      (e.g. "gmail", "slack", "gdrive"). Default is "unknown".
      add_document(doc_id: str, content: str, source: str = "unknown")

  search_by_source(keyword: str, source: str) -> list[str]
      Search filtered to a specific source. Sorted alphabetically.

Level 4 — Prefix Search / Autocomplete (15 min)
-------------------------------------------------
  autocomplete(prefix: str, limit: int = 5) -> list[str]
      Return unique words in the index that start with prefix
      (case-insensitive). Sorted alphabetically. Limited to
      `limit` results.

  get_index_stats() -> dict
      Return {"documents": int, "unique_words": int,
              "sources": list[str]}
      where sources is sorted alphabetically.

Examples
--------
  idx = SearchIndex()
  idx.add_document("d1", "hello world hello", source="gmail")
  idx.search("hello")           # -> ["d1"]
  idx.search_ranked("hello")    # -> ["d1"]  (2 occurrences)
  idx.autocomplete("hel")       # -> ["hello"]
"""


class SearchIndex:
    def __init__(self):
        # TODO: initialize your data structures
        pass

    # ── Level 1 ──────────────────────────────────────────────

    def add_document(self, doc_id: str, content: str, source: str = "unknown") -> None:
        """Index a document. Replace if doc_id already exists."""
        raise NotImplementedError("Level 1: add_document")

    def search(self, keyword: str) -> list[str]:
        """Return doc_ids containing keyword (case-insensitive), sorted."""
        raise NotImplementedError("Level 1: search")

    def remove_document(self, doc_id: str) -> bool:
        """Remove from index. Return False if not found."""
        raise NotImplementedError("Level 1: remove_document")

    # ── Level 2 ──────────────────────────────────────────────

    def search_all(self, keywords: list[str]) -> list[str]:
        """Return doc_ids containing ALL keywords (AND). Sorted."""
        raise NotImplementedError("Level 2: search_all")

    def search_any(self, keywords: list[str]) -> list[str]:
        """Return doc_ids containing ANY keyword (OR). Sorted."""
        raise NotImplementedError("Level 2: search_any")

    def get_word_count(self, doc_id: str) -> int:
        """Number of unique words in the document. -1 if not found."""
        raise NotImplementedError("Level 2: get_word_count")

    # ── Level 3 ──────────────────────────────────────────────

    def search_ranked(self, keyword: str) -> list[str]:
        """Doc_ids sorted by keyword frequency desc, then id asc."""
        raise NotImplementedError("Level 3: search_ranked")

    def search_by_source(self, keyword: str, source: str) -> list[str]:
        """Search filtered to a specific source. Sorted alphabetically."""
        raise NotImplementedError("Level 3: search_by_source")

    # ── Level 4 ──────────────────────────────────────────────

    def autocomplete(self, prefix: str, limit: int = 5) -> list[str]:
        """Words starting with prefix (case-insensitive), sorted, limited."""
        raise NotImplementedError("Level 4: autocomplete")

    def get_index_stats(self) -> dict:
        """Return {documents: int, unique_words: int, sources: list[str]}."""
        raise NotImplementedError("Level 4: get_index_stats")


# ─── Self-Checks (do not edit below this line) ──────────────────

_passed = 0
_failed = 0

def _check(label: str, actual: object, expected: object) -> None:
    global _passed, _failed
    if actual == expected:
        _passed += 1
        print(f"  \u2713 {label}")
    else:
        _failed += 1
        print(f"  \u2717 {label}")
        print(f"    expected: {expected!r}")
        print(f"         got: {actual!r}")

def _level(name: str, fn) -> None:
    global _failed
    print(name)
    try:
        fn()
    except NotImplementedError as e:
        print(f"  \u25cb {e}")
    except Exception as e:
        _failed += 1
        print(f"  \u2717 {e}")

def _run_self_checks() -> None:

    def level_1():
        idx = SearchIndex()
        # Basic add and search
        idx.add_document("doc1", "the quick brown fox")
        idx.add_document("doc2", "the lazy brown dog")
        idx.add_document("doc3", "quick fox jumps high")
        _check("search 'brown'",
               idx.search("brown"), ["doc1", "doc2"])
        _check("search 'quick'",
               idx.search("quick"), ["doc1", "doc3"])

        # Case-insensitive search
        idx.add_document("doc4", "Hello World HELLO")
        _check("case-insensitive search 'hello'",
               idx.search("hello"), ["doc4"])
        _check("case-insensitive search 'HELLO'",
               idx.search("HELLO"), ["doc4"])

        # Search miss
        _check("search miss",
               idx.search("zebra"), [])

        # Replace existing document
        idx.add_document("doc1", "completely new content")
        _check("replaced doc no longer matches 'fox'",
               idx.search("fox"), ["doc3"])
        _check("replaced doc matches 'completely'",
               idx.search("completely"), ["doc1"])

        # Remove
        _check("remove existing returns True",
               idx.remove_document("doc2"), True)
        _check("removed doc gone from search",
               idx.search("lazy"), [])
        _check("remove missing returns False",
               idx.remove_document("doc999"), False)

    def level_2():
        idx = SearchIndex()
        idx.add_document("d1", "python java rust")
        idx.add_document("d2", "python rust go")
        idx.add_document("d3", "java go elixir")
        idx.add_document("d4", "python java go rust elixir")

        # search_all — AND
        _check("search_all ['python', 'rust']",
               idx.search_all(["python", "rust"]),
               ["d1", "d2", "d4"])
        _check("search_all ['java', 'go', 'elixir']",
               idx.search_all(["java", "go", "elixir"]),
               ["d3", "d4"])
        _check("search_all no match",
               idx.search_all(["python", "elixir", "haskell"]),
               [])

        # search_any — OR
        _check("search_any ['elixir']",
               idx.search_any(["elixir"]),
               ["d3", "d4"])
        _check("search_any ['python', 'elixir']",
               idx.search_any(["python", "elixir"]),
               ["d1", "d2", "d3", "d4"])
        _check("search_any no match",
               idx.search_any(["haskell", "cobol"]),
               [])

        # get_word_count
        _check("word count d1", idx.get_word_count("d1"), 3)
        _check("word count d4", idx.get_word_count("d4"), 5)
        _check("word count missing", idx.get_word_count("d99"), -1)

        # Duplicate words — unique count
        idx.add_document("d5", "go go go python python")
        _check("word count with dupes", idx.get_word_count("d5"), 2)

    def level_3():
        idx = SearchIndex()
        # Frequency ranking: "data" appears 3x in d1, 1x in d2, 2x in d3
        idx.add_document("d1", "data data data science", source="gmail")
        idx.add_document("d2", "data engineering", source="slack")
        idx.add_document("d3", "big data data pipeline", source="gmail")
        idx.add_document("d4", "machine learning model", source="gdrive")

        _check("search_ranked by frequency",
               idx.search_ranked("data"),
               ["d1", "d3", "d2"])  # 3, 2, 1 occurrences

        # Tie-breaking by doc_id alphabetically
        idx.add_document("a_doc", "data results", source="slack")
        idx.add_document("z_doc", "data results", source="slack")
        ranked = idx.search_ranked("data")
        # a_doc and z_doc both have 1 occurrence, same as d2
        _check("search_ranked tie-breaking",
               ranked,
               ["d1", "d3", "a_doc", "d2", "z_doc"])

        _check("search_ranked no match",
               idx.search_ranked("quantum"), [])

        # search_by_source
        _check("search_by_source gmail 'data'",
               idx.search_by_source("data", "gmail"),
               ["d1", "d3"])
        _check("search_by_source slack 'data'",
               idx.search_by_source("data", "slack"),
               ["a_doc", "z_doc"])
        _check("search_by_source gdrive 'data'",
               idx.search_by_source("data", "gdrive"),
               [])

        # Case-insensitive ranked
        idx.add_document("d5", "Data DATA data", source="gmail")
        _check("ranked case-insensitive — d5 has 3 occurrences",
               idx.search_ranked("DATA")[0:2],
               ["d1", "d5"])  # both have 3

    def level_4():
        idx = SearchIndex()
        idx.add_document("d1", "dropbox desktop application sync", source="gdrive")
        idx.add_document("d2", "dropbox drive document sharing", source="gmail")
        idx.add_document("d3", "application programming interface api", source="slack")

        # Autocomplete basic
        _check("autocomplete 'drop'",
               idx.autocomplete("drop"),
               ["dropbox"])
        _check("autocomplete 'dr'",
               idx.autocomplete("dr"),
               ["drive", "dropbox"])
        _check("autocomplete 'app'",
               idx.autocomplete("app"),
               ["application"])

        # Autocomplete case-insensitive
        _check("autocomplete 'D' case-insensitive",
               idx.autocomplete("D"),
               ["desktop", "document", "drive", "dropbox"])

        # Autocomplete with limit
        idx.add_document("d4", "alpha beta gamma delta epsilon zeta")
        _check("autocomplete limit 3",
               idx.autocomplete("", 3),
               ["alpha", "api", "application"])

        # Autocomplete no match
        _check("autocomplete no match",
               idx.autocomplete("xyz"),
               [])

        # get_index_stats
        stats = idx.get_index_stats()
        _check("stats documents count",
               stats["documents"], 4)
        _check("stats unique_words > 0",
               stats["unique_words"] > 0, True)
        _check("stats sources sorted",
               stats["sources"],
               ["gdrive", "gmail", "slack"])

        # Remove a doc and check stats update
        idx.remove_document("d3")
        stats2 = idx.get_index_stats()
        _check("stats after remove — documents count",
               stats2["documents"], 3)

    _level("Level 1 \u2014 Basic Indexing", level_1)
    _level("Level 2 \u2014 Multi-Keyword Search", level_2)
    _level("Level 3 \u2014 Ranked Results", level_3)
    _level("Level 4 \u2014 Prefix Search / Autocomplete", level_4)

def main() -> None:
    print("\nDrill 06 \u2014 Search Index\n")
    _run_self_checks()
    total = _passed + _failed
    print(f"\n{_passed}/{total} passed")
    if _failed == 0 and total > 0:
        print("All tests passed.")

if __name__ == "__main__":
    main()
